src/server/twemcache-rs/src/admin.rs

```diff
@@ -0,0 +1,385 @@
+    addr: SocketAddr,
+    config: Arc<Config>,
+    listener: TcpListener,
+    poll: Poll,
+    ssl_context: Option<SslContext>,
+    sessions: Slab<Session>,
+    message_receiver: Receiver<Message>,
+    message_sender: Sender<Message>,
+}
+
+pub const LISTENER_TOKEN: usize = usize::MAX;
+
+impl Admin {
+    /// Creates a new `Admin` event loop.
+    pub fn new(config: Arc<Config>) -> Result<Self, Error> {
+        let addr = config.admin().socket_addr().map_err(|e| {
+            error!("{}", e);
+            std::io::Error::new(std::io::ErrorKind::Other, "Bad listen address")
+        })?;
+        let mut listener = TcpListener::bind(addr).map_err(|e| {
+            error!("{}", e);
+            std::io::Error::new(std::io::ErrorKind::Other, "Failed to start tcp listener")
+        })?;
+        let poll = Poll::new().map_err(|e| {
+            error!("{}", e);
+            std::io::Error::new(std::io::ErrorKind::Other, "Failed to create epoll instance")
+        })?;
+
+        // TODO(bmartin): does this need to be configurable to require TLS for
+        // admin?
+        // let ssl_context = crate::common::ssl_context(&config)?;
+        let ssl_context = None;
+
+        // register listener to event loop
+        poll.registry()
+            .register(&mut listener, Token(LISTENER_TOKEN), Interest::READABLE)
+            .map_err(|e| {
+                error!("{}", e);
+                std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "Failed to register listener with epoll",
+                )
+            })?;
+
+        let sessions = Slab::<Session>::new();
+
+        let (message_sender, message_receiver) = crossbeam_channel::bounded(128);
+
+        Ok(Self {
+            addr,
+            config,
+            listener,
+            poll,
+            ssl_context,
+            sessions,
+            message_sender,
+            message_receiver,
+        })
+    }
+
+    /// Runs the `Admin` in a loop, accepting new sessions for the admin
+    /// listener and handling events on existing sessions.
+    pub fn run(&mut self) {
+        info!("running admin on: {}", self.addr);
+
+        let mut events = Events::with_capacity(self.config.server().nevent());
+        let timeout = Some(std::time::Duration::from_millis(
+            self.config.server().timeout() as u64,
+        ));
+
+        // run in a loop, accepting new sessions and events on existing sessions
+        loop {
+            increment_counter!(&Stat::AdminEventLoop);
+
+            if self.poll.poll(&mut events, timeout).is_err() {
+                error!("Error polling");
+            }
+
+            increment_counter_by!(
+                &Stat::AdminEventTotal,
+                events.iter().count().try_into().unwrap(),
+            );
+
+            // handle all events
+            for event in events.iter() {
+                // handle new sessions
+                if event.token() == Token(LISTENER_TOKEN) {
```

Comment:
The inner loop is pretty long (ln 109-204), consider having the handling of events for new connection and on existing connection separated into their own routines for better readability.

Reply:
This should be better now.